Return `secret` from `UserVerificationDialogComponent().submit`

addisonbeck opened the pull request:
## Type of change

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

I'd like the ability to use `UserVerificationDialogComponent` for its reusable UI, but still build my own request for server side validation as part of a later API request. 

I'm seeking feedback from Auth on if they want their component to be able to do this, or if there is a better way I missed.

JaredSnider-Bitwarden commented on libs/auth/src/angular/user-verification/user-verification-dialog.types.ts:
```diff
@@ -60,12 +61,25 @@ export type UserVerificationDialogOptions = {
+   *   **Important:** This option is only for use on desktop and browser
+   *   platforms. When there are no client verification methods the user is
+   *   instructed to set a pin, and this is not supported on web.
+   * - "custom": Custom validation is done to verify the secret. This is
```
💡 could we expand the custom scenario comment to better explain the intended use case of providing a mechanism where users can call a secured endpoint which performs the user verification server side?

addisonbeck replied:
Implemented as suggested on https://github.com/bitwarden/clients/pull/8770/commits/0bf7129ac7f59d5d9a1a109650a8f43a90a3e9a9

JaredSnider-Bitwarden commented on apps/web/src/app/admin-console/organizations/users/enroll-master-password-reset.component.ts:
```diff
@@ -26,43 +28,49 @@ export class EnrollMasterPasswordReset {
+          );
+
+          // Process the enrollment request, which is an endpoint that is
+          // gated by a server-side check of the master password hash
```
Question: Would a TDE user without a MP ever hit this logic?  

Note:  new TDE users that are JIT provisioned on SSO login are automatically enrolled in admin account recovery so I'm not sure that they would.  

If they did for whatever reason, then the user verification dialog would present an OTP instead of a MP for them. Do we think that is a possibility to where we should change this comment to include OTP?

addisonbeck replied:
I feel like this is improbable enough not to be concerned with. This component is on accessible for users who are part of an organization before the account recovery policy is turned on, and so have a master password that was unmanaged by the organization. TDE shouldn't end up here, and if they do it would be indicative of a higher level issue.

JaredSnider-Bitwarden replied:
> This component is only accessible for users who are part of an organization without automatic account recovery enrollment turned on

That's a bit worrying b/c [our docs ](https://bitwarden.com/help/setup-sso-with-trusted-devices/) for TDE state that existing users must self enroll using this component before they can use the admin approval flows. Thus, we do expect users without MPs to end up on this component.

addisonbeck replied:
These users would still have a master password. The existing users in the use case outlined in that documentation would have had accounts provisioned and set up before any trusted device configuration was set up for the organization, and so would be using MP auth.

JaredSnider-Bitwarden replied:
Ah, I'm sorry. I was conflating having a MP vs having a local MP hash to compare against which isn't relevant in your scenario. I went and confirmed that, in the `UserVerificationService.getAvailableVerificationOptions(...)` method, we would always show the MP entry field if the user has a MP - regardless of local hash presence. 

To further clarify, this is the scenario I was trying to ensure wasn't an issue: 
If you have a user with a MP in an standard MP org that turns on TDE, they can login via SSO and approve via their MP + trust the device. Then, subsequent logins will decrypt using the device key - with no local MP hash to compare against for user verification scenarios.

JaredSnider-Bitwarden commented on apps/web/src/app/admin-console/organizations/users/enroll-master-password-reset.component.ts:
```diff
@@ -26,43 +28,49 @@ export class EnrollMasterPasswordReset {
         text: "resetPasswordEnrollmentWarning",
         type: "warning",
       },
+      verificationType: {
```
🏅 Thank you very much for the real scenario test.  Would you mind enshrining a condensed example in an example comment on the static `UserVerificationDialog.open` method so it's quite self explanatory to other devs that have a similar use case?

addisonbeck replied:
Added as suggested on https://github.com/bitwarden/clients/pull/8770/commits/ab41dbbf00ebdd89dd7c37325e33d1ed7d0efe79

JaredSnider-Bitwarden commented on libs/auth/src/angular/user-verification/user-verification-dialog.component.ts:
```diff
@@ -201,6 +201,17 @@ export class UserVerificationDialogComponent {
+      typeof this.dialogOptions.verificationType === "object" &&
+      this.dialogOptions.verificationType.type === "custom"
+    ) {
+      const success = await this.dialogOptions.verificationType.verificationFn(this.secret.value);
+      this.close({
+        userAction: "confirm",
+        verificationSuccess: success,
+      });
```
💡  @jlf0dev and I discussed an idea earlier of wrapping this function execution in a try catch which would attempt to handle all user verification errors in one location and it would give the user the chance to retry the validation.  All the server side validation is done via the `userService.verifySecret`, and almost all the errors are of the shape: `throw new BadRequestException(string.Empty, "User verification failed.");`

Thoughts?

addisonbeck replied:
This is a great idea, and works out well. I was actually able to just move the conditional that checks for custom validation a few lines down into the existing `try/catch` block.

JaredSnider-Bitwarden reviewed (commented):
This is looking really good so far; thank you for your work on this! A few questions and ideas to discuss below!

addisonbeck commented:
In addition to addressing the review feedback I have updated some new existing uses of `UserVerificationDialog` that use the `clientSideOnly` field that this PR replaces with `verificationType`. This adds a review dependency for vault team.

I've also added a screen recording of the account recovery enrollment use case in action.

JaredSnider-Bitwarden reviewed (dismissed):
🏅 Awesome job on these changes. Really like the new feature and thank you for the recording demoing it in action!

JaredSnider-Bitwarden commented on libs/auth/src/angular/user-verification/user-verification-dialog.types.ts:
```diff
@@ -60,12 +61,27 @@ export type UserVerificationDialogOptions = {
+   *
+   * Possible values:
+   *
+   * - "default": Perform the default validation operation for a given secret
+   *   type. This would, for example, validate master passwords locally but
+   *   OTPs on the server.
```
@eliykat , for context: [this PR](https://github.com/bitwarden/clients/pull/7536) explains the intent behind the component, links to the figmas, and provides videos of the user verification dialog in action. Please let me know if I can clarify anything!

eliykat replied:
Thank you both! Yes I think that was contributing tot he conclusion, in any case I'll also look into it separately.

addisonbeck commented on apps/web/src/app/admin-console/organizations/users/enroll-master-password-reset.component.ts:
```diff
@@ -26,43 +28,49 @@ export class EnrollMasterPasswordReset {
+        verificationFn: async (secret: VerificationWithSecret) => {
+          const request = (await userVerificationService.buildRequest(
+            secret,
+          )) as OrganizationUserResetPasswordEnrollmentRequest;
```
Implemented as suggested on https://github.com/bitwarden/clients/pull/8770/commits/0644431ffc8d9d0b96795d09b21890e56a2575fb

addisonbeck commented on libs/auth/src/angular/user-verification/user-verification-dialog.component.ts:
```diff
@@ -142,6 +142,33 @@ export class UserVerificationDialogComponent {
+   *     verificationFn: async (secret: VerificationWithSecret) => {
+   *       const request = (await userVerificationService.buildRequest(
+   *         secret,
+   *       )) as CustomRequestType;
```
Implemented as suggested on https://github.com/bitwarden/clients/pull/8770/commits/0644431ffc8d9d0b96795d09b21890e56a2575fb
